src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py

```diff
@@ -0,0 +1,410 @@
+        2. for every layer:
+            Choose Num Filters to prune
+            Choose which filter to prunee (l1)
+            Short-term fine tune the pruned model
+        3. Pick the best layer to prune
+    Long-term fine tune
+
+    For the details of this algorithm, please refer to the paper: https://arxiv.org/abs/1804.03230
+    """
+
+    def __init__(self, model, config_list, evaluator, fine_tuner,
+                 optimize_mode='maximize', pruning_mode='channel', pruning_step=0.05, experiment_data_dir='./'):
+        """
+        Parameters
+        ----------
+        model : pytorch model
+            The model to be pruned
+        config_list : list
+            Supported keys:
+                - sparsity : The final sparsity when the compression is done.
+                - op_types : The operation type to prune.
+        evaluator : function
+            function to evaluate the masked model
+        fine_tuner : function
+            function to short-term fine tune the masked model
+        optimize_mode : str
+            optimize mode, 'maximize' or 'minimize', by default 'maximize'
+        pruning_mode : str
+            'channel' or 'fine_grained', by default 'channel'
+        experiment_data_dir : str
+            PATH to save experiment data
+        """
+        # models used for iterative pruning and evaluation
+        self._model_to_prune = copy.deepcopy(model)
+        self._pruning_mode = pruning_mode
+
+        super().__init__(model, config_list)
+
+        self._fine_tuner = fine_tuner
+        self._evaluator = evaluator
+        self._optimize_mode = OptimizeMode(optimize_mode)
+
+        # hyper parameters for NetAdapt algorithm
+        self._pruning_step = pruning_step
+
+        # overall pruning rate
+        self._sparsity = config_list[0]['sparsity']
+
+        # config_list
+        self._config_list = []
+
+        self._experiment_data_dir = experiment_data_dir
+        if not os.path.exists(self._experiment_data_dir):
+            os.makedirs(self._experiment_data_dir)
+
+        self._tmp_model_path = os.path.join(
+            self._experiment_data_dir, 'tmp_model.pth')
+
+        self._total_num_weights = self._get_total_num_weights()
+
+    def _detect_modules_to_compress(self):
+        """
+        redefine this function, consider only the layers without dependencies
+        """
+        self._op_names_to_compress = []
+        if self.modules_to_compress is None:
+            self.modules_to_compress = []
+            # consider only the layers without dependencies
+            model_name = self._model_to_prune.__class__.__name__
+            ops_no_dependency = get_layers_no_dependency(model_name)
+
+            for name, module in self.bound_model.named_modules():
+                if module == self.bound_model:
+                    continue
+                if self._pruning_mode == 'channel' and model_name in ['MobileNetV2', 'RetinaFace'] and name not in ops_no_dependency:
```

Comment:
This checking seems very specific, would you explain the reason to do this check?

Reply:
The PR https://github.com/microsoft/nni/pull/2435 provides functionalities to check dependencies. The checking here will be replaced after this PR is merged.

Reply:
The checking is removed. To make it clear, I'll deal with this functionality (prune only ops without dependency) in a seperate PR in the future.